mobilecoind/src/conversions.rs

```diff
@@ -315,6 +365,84 @@ mod test {
+            &RistrettoPrivate::from_random(&mut rng),
+            &RistrettoPrivate::from_random(&mut rng),
+        );
+        let alice_cred = SenderMemoCredential::from(&alice);
+        let alice_hash = alice_cred.address_hash.clone();
+
+        let bob = AccountKey::new(
+            &RistrettoPrivate::from_random(&mut rng),
+            &RistrettoPrivate::from_random(&mut rng),
+        );
+        let bob_addr = bob.default_subaddress();
+
+        let tx_public_key = CompressedRistrettoPublic::from_random(&mut rng);
+
+        let memo1 = UnusedMemo {};
+        let decoded = decode_memo(&MemoPayload::from(memo1));
+        assert!(!decoded.has_authenticated_sender_memo());
+        assert!(!decoded.has_unknown_memo());
+
+        let memo2 =
+            AuthenticatedSenderMemo::new(&alice_cred, bob_addr.view_public_key(), &tx_public_key);
+        let decoded = decode_memo(&MemoPayload::from(memo2));
+        assert!(decoded.has_authenticated_sender_memo());
+        assert!(!decoded.has_unknown_memo());
+        let sender_memo = decoded.get_authenticated_sender_memo();
+        assert_eq!(sender_memo.get_sender_hash(), alice_hash.as_ref());
+        assert!(!sender_memo.has_payment_request_id());
+        assert!(!sender_memo.has_payment_intent_id());
+
+        let memo3 = AuthenticatedSenderWithPaymentRequestIdMemo::new(
+            &alice_cred,
+            bob_addr.view_public_key(),
+            &tx_public_key,
+            7u64,
+        );
+        let decoded = decode_memo(&MemoPayload::from(memo3));
+        assert!(decoded.has_authenticated_sender_memo());
+        assert!(!decoded.has_unknown_memo());
+        let sender_memo = decoded.get_authenticated_sender_memo();
+        assert_eq!(sender_memo.get_sender_hash(), alice_hash.as_ref());
+        assert!(sender_memo.has_payment_request_id());
+        assert_eq!(sender_memo.get_payment_request_id(), 7);
+        assert!(!sender_memo.has_payment_intent_id());
+
+        let memo4 = AuthenticatedSenderWithPaymentIntentIdMemo::new(
+            &alice_cred,
+            bob_addr.view_public_key(),
+            &tx_public_key,
+            9u64,
+        );
+        let decoded = decode_memo(&MemoPayload::from(memo4));
+        assert!(decoded.has_authenticated_sender_memo());
+        assert!(!decoded.has_unknown_memo());
+        let sender_memo = decoded.get_authenticated_sender_memo();
+        assert_eq!(sender_memo.get_sender_hash(), alice_hash.as_ref());
+        assert!(!sender_memo.has_payment_request_id());
+        assert!(sender_memo.has_payment_intent_id());
+        assert_eq!(sender_memo.get_payment_intent_id(), 9);
+
+        // Destination memos are not implemented yet
+        let memo5 = DestinationMemo::new(ShortAddressHash::from(&bob_addr), 17, 18).unwrap();
+        let decoded = decode_memo(&MemoPayload::from(memo5));
+        assert!(!decoded.has_authenticated_sender_memo());
+        assert!(decoded.has_unknown_memo());
```

Comment:
Recommend asserting that `type_bytes` are properly set

Reply:
:100: